### ISSUE TITLE
Multithreading support

### ISSUE THREAD

**Issue by chrispytoes** [Quillraven/Fleks#82]
I was wondering about the possibility of multithreading support by making use of the fact that not all systems will depend on the same entities, so why not allow unrelated systems to run in parallel on separate threads?

It could look something like this:
```
world {
  systems {
    add(FirstSystem())
    parallel {
      add(ParallelSystem1())
      add(ParallelSystem2())
    }
    add(FinalSystem())
  }
}
```

Of course, you would have to use this at your own digression and make sure they don't share any dependencies, unless it can be auto-calculated by comparing their families and injectables.

Is something like this achievable in the current architecture?

**Comment by Quillraven** [comment]
This was a very early decision in the project to not try to make it multi-threaded because the complexity grows a lot in my opinion.

Currently it won't work because a lot of internal things are not threadsafe like families, BitArray, EntityBag, etc...

Also, imo you cannot really run everything in separate threads when it comes to game logic. For me it makes sense to e.g. run audio and rendering separately but the logic itself must be in a sequence because otherwise you simply get a mess? Or how would it then work e.g. in a RPG where you can move/take damage/deal damage/cast spells, etc. and everything is happening in parallel meaning that my caster unit can die while it maybe also casted a spell at the same time. 

Also, how would family iteration work if your two parallel systems are iterating over the same family (or similar family) and you remove entities during those iterations?

So in short: sorry, but currently that is not supported and I don't have plans to do that. BUT of course you are more than welcome  to make a PR to add it ;)

However, what you can do is of course to make one of your systems use coroutines. E.g. you create your own `ParallelSystem` and override the `onUpdate` function and run your parallel logic in this function. I could imagine a `runBlocking` to make sure that everything finishes at the end of the `onUpdate` function. That way you can still use coroutines but it is then your responsiblity to make sure that everything is running properly. What do you think?

**Comment by chrispytoes** [comment]
@Quillraven Yeah I suppose I asked this question a little too hastily, you could just use coroutines within a system. I suppose I'll just run my Fleks world and renderer in separate threads and use a `Channel` to pass messages between them. Then if I need any more extra performance I can always use coroutines in a system.

**Comment by Quillraven** [comment]
Don't get me wrong. Official coroutine support in ECS would be great but I don't have that much experience yet with coroutines and I just know from Java threads that it is sometimes not so easy to keep your codebase threadsafe. That's why currently it is not supported by Fleks, sorry.

If you have an example of a system that runs some code via coroutines I'd be interested. Also about your renderer and channel communication. Would be great if you can share your solution 😊

**Comment by jobe-m** [comment]
Very interesting topic BTW
I am using fleks with korge. And I face exactly that problems in putting both together. Fleks is very deterministic - single threaded while korge is highly dynamic. Basically everything can run in parallel. (Korge has coroutines in is name;) i.e. when using fleks for some specific task I need to "switch that off" in korge. E.g. I wasn't satisfied with using the animation system of korge to animate properties of components and ended up writing my own animation system in fleks.

Currently my integration feels like two worlds living side by side ;)

**Comment by chrispytoes** [comment]
@Quillraven I very well might, I'm actually just learning Kotlin for the first time with this project but I come from a background in JS and Go where parallelism is just the natural order of things, so it just felt right to include it in here too. For now, I can leave you with this system I came up with for running multiple systems in parallel and waiting for all of them:
```
class ParallelSystem(
    private val systems: List<IntervalSystem>,
): IntervalSystem(
    interval = EachFrame
) {
    private val threads: List<Thread> = systems.map {
        Thread {
            it.onUpdate()
        }
    }

    override fun onTick() {
        threads.forEach {
            it.start()
        }

        threads.forEach {
            it.join()
        }
    }
}
```
This works, but of course you can only add one of them to your world since it restricts to only one instance of each system, but you could just copy/paste this pattern into other systems instead.

@jobe-m I'm actually using Korge as well, and yeah I love how dynamic it is, and yeah it has great parallelism, but not real threads. If you try to call `View`s from multiple threads with it it will give you a lot of errors really quick. I've always really wished there were game engines with no graphics, just all the object management and everything else but let me do the rendering somewhere else. Large-scale multithreaded simulations would be much easier that way.
